src/Features/CSharp/Portable/UseRecursivePatterns/CSharpUseRecursivePatternsCodeRefactoringProvider.AnalyzedNode.cs

```diff
@@ -0,0 +1,277 @@
+            Conjunction = 1,
+            VarPattern = 1 << 1,
+            TypePattern = 1 << 2,
+            NotNullPattern = 1 << 3,
+            ConstantPattern = 1 << 4,
+            PatternMatch = 1 << 5,
+            Evaluation = 1 << 6,
+            PositionalPattern = 1 << 7,
+            DiscardPattern = 1 << 8,
+        }
+
+        private abstract class AnalyzedNode
+        {
+            public abstract NodeKind Kind { get; }
+
+            public virtual AnalyzedNode Reduce(Reducer reducer) => this;
+            public virtual bool Contains(ExpressionSyntax e) => false;
+            public virtual void GetChildren(List<AnalyzedNode> nodes) => nodes.Add(this);
+
+            public virtual ExpressionSyntax AsExpressionSyntax() => throw ExceptionUtilities.UnexpectedValue(this);
+            public virtual PatternSyntax AsPatternSyntax() => throw ExceptionUtilities.UnexpectedValue(this);
+
+            public abstract override string ToString();
+        }
+
+        private sealed class Conjunction : AnalyzedNode
+        {
+            public readonly AnalyzedNode Left;
+            public readonly AnalyzedNode Right;
+
+            public override NodeKind Kind => NodeKind.Conjunction;
+
+            public static Conjunction Create(AnalyzedNode left, AnalyzedNode right)
+            {
+                if (left is null || right is null)
+                    return null;
+
+                return new Conjunction(left, right);
+            }
+
+            public Conjunction(AnalyzedNode left, AnalyzedNode right)
+            {
+                Debug.Assert(left != null);
+                Debug.Assert(right != null);
+                Left = left;
+                Right = right;
+            }
+
+            public override string ToString() => $"{Left} AND {Right}";
+
+            public override bool Contains(ExpressionSyntax e) => Left.Contains(e) || Right.Contains(e);
+
+            public override AnalyzedNode Reduce(Reducer reducer) => reducer.ReduceConjunction(this);
+
+            public override void GetChildren(List<AnalyzedNode> nodes)
+            {
+                Left.GetChildren(nodes);
+                Right.GetChildren(nodes);
+            }
+
+            public override PatternSyntax AsPatternSyntax() => AsPatternSyntax(out _);
+
+            private PatternSyntax AsPatternSyntax(out List<AnalyzedNode> children)
+            {
+                children = new List<AnalyzedNode>();
+                GetChildren(children);
+                return RecursivePattern(
+                    children.OfType<TypePattern>().SingleOrDefault()?.Type,
+                    children.OfType<PositionalPattern>().SingleOrDefault()?.AsPositionalPatternClauseSyntax(),
+                    PropertyPatternClause(SeparatedList(
+                        children.OfType<PatternMatch>().Select(match => match.AsSubpatternSyntax()))),
+                    children.OfType<VarPattern>().SingleOrDefault()?.AsVariableDesignationSyntax());
```

Comment:
doc all of this.

Reply:
I think I should try to comment the high level idea here, cause comments for each of these will be a prose of the same line.